book/src/programs.md

```diff
@@ -1,32 +1,46 @@
-written in the client’s choice of programming language.
+The Solana *runtime* can execute untrusted on-chain programs written in any
+programming language that can target the Berkley Packet Filter (BPF) safe
+execution environment. The Solana SDK offers the best support for C programs,
+which is compiled to BPF using the [LLVM compiler
+infrastructure](https://llvm.org). Alternatively, a client might choose to
+bypass LLVM and use Python, Lua or C++ to generate BPF directly via the [BPF
+Compiler Collection](https://github.com/iovisor/bcc) (BCC).
 
-## Client Interactions with Solana
+## Deploying Programs to a Cluster
 
 <img alt="SDK tools" src="img/sdk-tools.svg" class="center"/>
 
-As shown in the diagram above an untrusted client, creates a program in the
-language of their choice, (i.e. C/C++/Rust/Lua), and compiles it with LLVM to a
-position independent shared object ELF, targeting BPF bytecode, and sends it to
-the Solana cluster. Next, the client sends messages to the Solana cluster,
-which target that program. The Solana runtime loads the previously submitted
-ELF and passes it the client's message for interpretation.
+As shown in the diagram above a client creates a program and compiles it to an
+ELF shared object containing BPF bytecode and sends it to the Solana cluster.
+The cluster stores the program locally and makes it available to clients via a
+*program ID*. The program ID is a public key generated by the client and is
+used to reference the program in subsequent transactions.
 
-## Persistent Storage
+## Interacting with On-chain Programs
 
-Solana supports several kinds of persistent storage, called *accounts*:
+After a program has been deployed, a client utilizes it by sending transactions
+to the Solana cluster. A transaction contains one or more *instructions*. Each
+instruction references the program ID that coresponds to the program that the
+runtime should use to execute it. Instructions are executed atomically. If any
+instruction is invalid, any changes made within the transaction are discarded.
+
+## Storing State between Transactions
+
+If the the program needs to store state between transactions, it does so using
+*accounts*. Solana supports several kinds of accounts:
 
 1. Executable
 2. Owned by a client
 3. Owned by a program
 4. Credit-only
 
-All accounts are identified by public keys and may hold arbirary data.
-When the client sends messages to programs, it requests access to storage
-using those keys. The runtime loads the account data and passes it to the
-program. The runtime also ensures accounts aren't written to if not owned
-by the client or program. Any writes to credit-only accounts are discarded
-unless the write was to credit tokens. Any user may credit other accounts
-tokens, regardless of account permission.
+All accounts are identified by public keys and may hold arbitrary data.  When
+the client sends messages to programs, it requests access to accounts using
```

Comment:
The wording of “sends messages” feels a little strange.  To this point the text has been discussing transactions and the instructions they contain, but doesnt define what a _message_ is

Reply:
Yeah, that was bugging me too. I'll change it.